src/librustc_mir/transform/add_yk_swt_calls.rs

```diff
@@ -0,0 +1,170 @@
+        let num_orig_local_decls = mir.local_decls.len();
+        let local_crate_hash = tcx.crate_hash(LOCAL_CRATE).as_u64();
+
+        for (bb, bb_data) in mir.basic_blocks_mut().iter_enumerated() {
+            // Copy the original block and compute what its index will be once we have pushed onto
+            // the end of the MIR's basic block vector.
+            let new_blk = bb_data.clone();
+            let new_blk_idx = BasicBlock::new(num_orig_blocks + copied_blocks.len());
+            copied_blocks.push(new_blk);
+
+            // Prepare to call the recorder function.
+            let ret_val = LocalDecl::new_temp(unit_ty, DUMMY_SP);
+            let ret_place = Place::Local(Local::new(num_orig_local_decls + new_local_decls.len()));
+            new_local_decls.push(ret_val);
+
+            let crate_hash_oper = Operand::Constant(box Constant {
+                span: DUMMY_SP,
+                ty: u64_ty,
+                user_ty: None,
+                literal: ty::Const::from_u64(tcx, local_crate_hash),
+            });
+
+            let def_idx_oper = Operand::Constant(box Constant {
+                span: DUMMY_SP,
+                ty: u32_ty,
+                user_ty: None,
+                literal: ty::Const::from_u32(tcx, self.0.as_raw_u32()),
+            });
+
+            let bb_oper = Operand::Constant(box Constant {
+                span: DUMMY_SP,
+                ty: u32_ty,
+                user_ty: None,
+                literal: ty::Const::from_u32(tcx, bb.index() as u32),
+            });
+
+            let rec_fn_oper = Operand::function_handle(tcx, rec_fn_defid,
+                List::empty(), DUMMY_SP);
+
+            let term_kind = TerminatorKind::Call {
+                func: rec_fn_oper,
+                args: vec![crate_hash_oper, def_idx_oper, bb_oper],
+                destination: Some((ret_place, new_blk_idx)),
+                cleanup: None,
+                from_hir_call: false,
+            };
+
+            // Construct a new block to replace the original one.
+            let source_info = bb_data.terminator.clone().map(|t| t.source_info)
+                .or(Some(SourceInfo { span: DUMMY_SP, scope: OUTERMOST_SOURCE_SCOPE })).unwrap();
+            let replace_block = BasicBlockData {
+                statements: vec![],
+                terminator: Some(Terminator {
+                    source_info,
+                    kind: term_kind
+                }),
+                is_cleanup: false
+            };
+            replace_blocks.push(replace_block);
+        }
+
+        // Finally, commit our transformations.
+        mir.basic_blocks_mut().extend(copied_blocks);
+        mir.local_decls.extend(new_local_decls);
+        for (bb, bb_data) in replace_blocks.drain(..).enumerate() {
+            mir.basic_blocks_mut()[BasicBlock::new(bb)] = bb_data;
+        }
+    }
+}
+
+/// Given a `MirSource`, decides if we should annotate the correpsonding MIR.
```

Comment:
Should this be "Given a `MirSource`, return `true` if the corresponding MIR should be annotated or false otherwise." Although I don't think "annotate" is the right word here: we really mean something like "should_be_traced"?

Reply:
> Should this be

It should be clear from the name of the function that true means we should.

> annotate vs. should_be_traced

We are annotating blocks with calls, right?

I'm not sure about `should_be_traced`. How about `should_transform`?

Reply:
You'd be amazed at how many functions that return bool aren't clear about what true/false means, so I find it best to be explicit. It's a bit like axes on graphs: often stating "higher is better" (or "lower is better") massively helps readers.

I'm actually wondering if we've got this function the wrong way around (so to speak). Maybe instead of "this can be annotated/transformed", it would be clearer to have the function be "can't be annotated/transformed"? Then the function might be called something like `is_untraceable`.